pkg/bff/organization.go

```diff
@@ -169,29 +171,64 @@ func (s *Server) ListOrganizations(c *gin.Context) {
 		return
 	}
 
+	// Parse the params from the GET request
+	params := &api.ListOrganizationsParams{}
+	if err = c.ShouldBindQuery(params); err != nil {
+		log.Warn().Err(err).Msg("could not bind request with query params")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		return
+	}
+
+	// Set pagination defaults
+	if params.Page <= 0 {
+		params.Page = 1
+	}
+
+	if params.PageSize <= 0 {
+		params.PageSize = 8
+	}
+
+	// Determine pagination index range (indexed by 1)
+	minIndex := (params.Page - 1) * params.PageSize
+	maxIndex := minIndex + params.PageSize
+	log.Debug().Int("page", params.Page).Int("page_size", params.PageSize).Int("min_index", minIndex).Int("max_index", maxIndex).Msg("paginating organizations")
+
 	// Build the response
-	out := make([]*api.OrganizationReply, 0)
+	out := &api.ListOrganizationsReply{
+		Organizations: make([]*api.OrganizationReply, 0),
+		Page:          params.Page,
+		PageSize:      params.PageSize,
+	}
+
+	// Iterate over the user's organization assignments and retrieve each organization
 	for _, id := range appdata.GetOrganizations() {
-		if org, err := s.OrganizationFromID(id); err != nil {
+		var org *models.Organization
+		if org, err = s.OrganizationFromID(id); err != nil {
 			log.Error().Err(err).Str("org_id", id).Msg("could not retrieve organization from database")
-		} else {
-			// User data is on the collaborator record.
-			// Note: The UserInfo middleware ensures that the user email address is
-			// present in the claims so we can safely dereference it here.
-			var collaborator *models.Collaborator
-			if collaborator = org.GetCollaborator(*user.Email); collaborator == nil {
-				log.Error().Str("org_id", id).Str("email", *user.Email).Msg("could not find user in organization collaborators")
-				continue
-			}
-
-			out = append(out, &api.OrganizationReply{
+			continue
+		}
+
+		// Make sure the user is a collaborator in the organization
+		// Note: The UserInfo middleware ensures that the user email address is
+		// present in the claims so we can safely dereference it here.
+		var collaborator *models.Collaborator
+		if collaborator = org.GetCollaborator(*user.Email); collaborator == nil {
+			log.Error().Str("org_id", id).Str("email", *user.Email).Msg("could not find user in organization collaborators")
+			continue
+		}
+
+		if out.Count >= minIndex && out.Count < maxIndex {
```

Comment:
Am I right in thinking that for each call to `ListOrganizations` you're iterating through all organizations? Wouldn't it be more efficient to only iterate through the number of organizations in `params.PageSize`?

Reply:
You're right the endpoint would be faster. We could design the pagination so that we only iterate up to the page limit as you suggest, the advantage of this approach is that returning the total count allows the client to compute how many pages there are for frontend purposes.